File: src/main/java/io/github/ytg1234/manhunt/command/SpeedrunnerCommand.java
package io.github.ytg1234.manhunt.command;

import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.context.CommandContext;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import io.github.ytg1234.manhunt.ManhuntUtils;
import net.minecraft.command.argument.EntityArgumentType;
import net.minecraft.server.command.ServerCommandSource;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.text.LiteralText;
import net.minecraft.text.TranslatableText;

import static net.minecraft.server.command.CommandManager.argument;
import static net.minecraft.server.command.CommandManager.literal;

public class SpeedrunnerCommand {
    public static void register(CommandDispatcher<ServerCommandSource> dispatcher) {
        dispatcher.register(literal("speedrunner").requires(source -> source.hasPermissionLevel(2))
                                                  .then(literal("set").then(argument("target", EntityArgumentType.player()).executes(
                                                          SpeedrunnerCommand::executeSet)))
                                                  .then(literal("get").executes(SpeedrunnerCommand::executeGet))
                                                  .then(literal("clear").executes(SpeedrunnerCommand::executeClear)));
    }


    private static int executeSet(CommandContext<ServerCommandSource> context) throws CommandSyntaxException {
        boolean playerHasMod = ManhuntUtils.playerHasMod(context);
        ServerPlayerEntity target = EntityArgumentType.getPlayer(context, "target");

        if (ManhuntUtils.hunters.contains(target.getUuid())) {
            if (playerHasMod) {
                context.getSource().sendError(new TranslatableText("text.manhunt.command.speedrunner.error.hunter", target.getDisplayName()));
            } else {
                context.getSource()
                       .sendError(new LiteralText("Cannot set speedrunner to ").append(target.getDisplayName())
                                                                               .append(new LiteralText(" because they are a hunter!")));
            }
            return 1;
        }

        ManhuntUtils.speedrunner = target.getUuid();
        if (playerHasMod) {
            context.getSource()
                   .sendFeedback(new TranslatableText("text.manhunt.command.speedrunner.set",
                                                      ManhuntUtils.fromCmdContext(context, ManhuntUtils.speedrunner).getDisplayName()
                   ), true);
        } else {
            context.getSource()
                   .sendFeedback(new LiteralText("Set speedrunner to ").append(target.getDisplayName()).append(new LiteralText("!")), true);
        }
        return 1;
    }

    private static int executeGet(CommandContext<ServerCommandSource> context) throws CommandSyntaxException {
        boolean playerHasMod = ManhuntUtils.playerHasMod(context);
        if (ManhuntUtils.speedrunner == null) return 1;

        if (playerHasMod) {
            context.getSource()
                   .sendFeedback(new TranslatableText("text.manhunt.command.speedrunner.get",
                                                      ManhuntUtils.fromCmdContext(context, ManhuntUtils.speedrunner).getDisplayName()
                   ), false);
        } else {
            context.getSource()
                   .sendFeedback(new LiteralText("Speedrunner is currently: ").append(ManhuntUtils.fromCmdContext(context, ManhuntUtils.speedrunner)
                                                                                                  .getDisplayName()), true);
        }
        return 1;
    }

    private static int executeClear(CommandContext<ServerCommandSource> context) throws CommandSyntaxException {
        boolean playerHasMod = ManhuntUtils.playerHasMod(context);
        ManhuntUtils.speedrunner = null;
        if (playerHasMod) {
            context.getSource().sendFeedback(new TranslatableText("text.manhunt.command.speedrunner.clear"), true);
        } else {
            context.getSource().sendFeedback(new LiteralText("Speedrunner Cleared!"), true);
        }
        return 1;
    }
}

File: manhunt-base/src/main/kotlin/io/github/ytg1234/manhunt/command/SpeedrunnerCommand.kt
package io.github.ytg1234.manhunt.command

import com.mojang.brigadier.Command
import com.mojang.brigadier.context.CommandContext
import io.github.ytg1234.manhunt.base.fromCmdContext
import io.github.ytg1234.manhunt.base.hunters
import io.github.ytg1234.manhunt.base.playerHasMod
import io.github.ytg1234.manhunt.base.speedrunner
import io.github.ytg1234.manhunt.util.PermedCommand
import io.github.ytg1234.manhunt.util.plus
import io.github.ytg1234.manhunt.util.reset
import net.minecraft.command.argument.EntityArgumentType
import net.minecraft.server.command.CommandManager
import net.minecraft.server.command.ServerCommandSource
import net.minecraft.text.TranslatableText

/**
 * Used to control the [speedrunner].
 *
 * @author YTG1234
 */
object SpeedrunnerCommand : PermedCommand("manhunt.command.speedrunner", 2) {
    override val cmd = CommandManager.literal("speedrunner").then(
        CommandManager.literal("set").then(CommandManager.argument("target", EntityArgumentType.player()).executes(::executeSet))
    ).then(
        CommandManager.literal("get").executes(::executeGet)
    ).then(
        CommandManager.literal("clear").executes(::executeClear)
    )

    /**
     * Changes the [speedrunner] to another player.
     */
    private fun executeSet(context: CommandContext<ServerCommandSource>): Int {
        val playerHasMod: Boolean = playerHasMod(context)
        val target = EntityArgumentType.getPlayer(context, "target")
        if (hunters.contains(target.uuid)) {
            if (playerHasMod) {
                context.source.sendError(
                    TranslatableText(
                        "text.manhunt.command.speedrunner.error.hunter",
                        target.displayName
                    )
                )
            } else {
                context.source
                    .sendError(
                        reset("Cannot set speedrunner to ") + target.displayName + reset(" because they are a hunter!")
                    )
            }
            return Command.SINGLE_SUCCESS
        }
        speedrunner = target.uuid
        if (playerHasMod) {
            context.source
                .sendFeedback(
                    TranslatableText(
                        "text.manhunt.command.speedrunner.set",
                        fromCmdContext(context, speedrunner)!!.displayName
                    ), true
                )
        } else {
            context.source
                .sendFeedback(
                    reset("Set speedrunner to ") + target.displayName + reset("!"),
                    true
                )
        }
        return Command.SINGLE_SUCCESS
    }

    /**
     * Sends the [speedrunner]'s name.
     */
    private fun executeGet(context: CommandContext<ServerCommandSource>): Int {
        val playerHasMod: Boolean = playerHasMod(context)
        if (speedrunner == null) return 1
        if (playerHasMod) {
            context.source
                .sendFeedback(
                    TranslatableText(
                        "text.manhunt.command.speedrunner.get",
                        fromCmdContext(context, speedrunner)!!.displayName
                    ), false
                )
        } else {
            context.source
                .sendFeedback(
                    reset("Speedrunner is currently: ") + fromCmdContext(context, speedrunner)!!.displayName,
                    true
                )
        }
        return Command.SINGLE_SUCCESS
    }

    /**
     * Sets the [speedrunner] to `null`
     */
    private fun executeClear(context: CommandContext<ServerCommandSource>): Int {
        val playerHasMod: Boolean = playerHasMod(context)
        speedrunner = null
        if (playerHasMod) {
            context.source.sendFeedback(TranslatableText("text.manhunt.command.speedrunner.clear"), true)
        } else {
            context.source.sendFeedback(reset("Speedrunner Cleared!"), true)
        }
        return Command.SINGLE_SUCCESS
    }
}
